### PR TITLE
[SPARK-13236] SQL Generation for Set Operations

### PR DESCRIPTION
**Pull request by gatorsmile**
This PR is to implement SQL generation for the following three set operations:
- Union Distinct
- Intersect
- Except

@liancheng Thanks!

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala [hunk 1]
```diff
@@ -105,6 +105,18 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
     checkHiveQl("SELECT id FROM t0 UNION ALL SELECT CAST(id AS INT) AS id FROM t0")
   }
 
+  test("union distinct") {
+    checkHiveQl("SELECT * FROM t0 UNION SELECT * FROM t0")
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: does this work if the query looks like

select * from a union (select * from b limit 10)?

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: Hive parser is unable to recognize the SQL. 

Do you want me to reopen this PR https://github.com/apache/spark/pull/10689? That is for resolving this issue. Thanks!

Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: well my worry is that if we do support that in the future, nobody will remember coming back to this place and fix it. let's just add parentheses around so it is robust against future changes.

Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: basically for the sql generation stuff, we need to anticipate future changes, and can't focus on what we have at this moment.

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: I see. Will do it. Thanks!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala [hunk 2]
```diff
@@ -105,6 +105,27 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
...
+  }
+
+  // Parser is unable to parse the following query:
+  // SELECT  `u_1`.`id`
+  // FROM (((SELECT  `t0`.`id` FROM `default`.`t0`)
+  // UNION ALL (SELECT  `t0`.`id` FROM `default`.`t0`))
+  // UNION ALL (SELECT  `t0`.`id` FROM `default`.`t0`)) AS u_1
+  ignore("three-child union") {
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: @hvanhovell @viirya Could you take a look at this issue? I am wondering if we need to fix this issue in the parser. Thanks!

Comment (hvanhovell) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: @gatorsmile Yeah we should fix this. The parser does not like the nesting within the UNION ALL. It parses fine if you remove the first nested `UNION ALL`:

    SELECT  `u_1`.`id`
    FROM ((SELECT  `t0`.`id` FROM `default`.`t0`)
    UNION ALL (SELECT  `t0`.`id` FROM `default`.`t0`)
    UNION ALL (SELECT  `t0`.`id` FROM `default`.`t0`)) AS u_1

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: Thank you! @hvanhovell 

Based on the comment by @rxin , I think we still need a nesting version here for future extension, instead hacking the SQL generation. 

Saw @viirya did a similar work last month: https://github.com/apache/spark/pull/10660. Maybe @viirya can do me a favor for completing this? Thank you! : )

Also cc @liancheng

Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: Yeah, I will look into this issue today.

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: Thank you! @viirya

Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: Submitted #11204 for this issue.

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/LogicalPlanToSQLSuite.scala]: Thank you for your fast work! @viirya

**Comment by gatorsmile** [comment]
ok to test

**Comment by liancheng** [comment]
LGTM, merging to master. Thanks!

**Pull request closed without merging**